python_modules/dagster/dagster/_core/pipes/merge_streams.py

```diff
@@ -0,0 +1,232 @@
+        except Exception as e:
+            logger.warning(f"An error occurred during processing '{entry.log}': {e}")
+        finally:
+            logs.task_done()
+
+
+@contextlib.contextmanager
+def merge_streams(
+    streams: Mapping[str, Generator],
+    log_handler: Callable,
+    stream_processor: Callable[[Iterator[bytes]], Iterator[LogItem]],
+    recent_messages_buffer_size: int = 500,
+    logger: logging.Logger = _default_logger,
+) -> Generator[None, None, None]:
+    """merge_streams and handle them with a single handler.
+
+    The list of streams will be processed in separate threads and the handler will accept
+    each message separately. We expect streams to produce log messages in the format of
+        b'<timestamp> <msg>'
+
+    Args:
+    ----
+        streams: A dict of generators that will provide log messages. It is expected to sometimes observe duplicate
+            messages, which will be deduplicated. The keys to the dictionary are the labels of each stream for
+            debugging purposes.
+        log_handler: The handler to use on each log message returned
+            by the stream.
+        stream_processor: The function that parses the streams into log lines.
+        recent_messages_buffer_size: The size of the buffer for storing the recent messages.
+        logger: The function for logging.
+
+    """
+    workers = queue.Queue()
+    logs = queue.PriorityQueue()
+
+    for label, log_stream in streams.items():
+        workers.put((label, log_stream))
+
+    shutdown = threading.Event()
+    threads = [
+        # Make threads the individual extract logs calls
+        safe_thread(
+            target=_enqueue,
+            kwargs={
+                "streams": workers,
+                "out": logs,
+                "recent_messages_buffer_size": recent_messages_buffer_size,
+                "stream_processor": stream_processor,
+                "logger": logger,
+            },
+            logger=logger,
+        )
+        for i in range(len(streams))
+    ] + [
+        # TODO @aignas 2024-03-22: we could add a buffer if there is a lot of lag between
+        # when the logs appear in the priority queue and when they are processed.
+        safe_thread(
+            target=_handle,
+            kwargs={
+                "logs": logs,
+                "log_handler": log_handler,
+                "shutdown": shutdown,
+                "interval": 1,
+                "logger": logger,
+            },
+            logger=logger,
+        ),
+    ]
+
+    for t in threads:
+        t.start()
+
+    try:
+        yield
+    finally:
+        # The shutdown process is:
+        # 1. Wait until all log produces are done
+        # 2. Add a special log item that will be processed the last and the handler thread will exit itself
+        # 3. Set the shutdown event to signal the thread that it should shutdown.
+        # 4. Join the queue
+        workers.join()
+        logs.put(
```

Comment:
Note, if you had multiple sinks, you would have to do the `logs.put("end-of-queue")` for each sink, however, the same `shutdown` threading event could be reused.